### PR TITLE
feat(log): add critical level to be used for alerting

### PR DESCRIPTION
**Pull request by marregui**
Adds `LogLevel.CRITICAL`, which is used to flag log messages as alert-able (the LogAlertSocketWriter reacts to these only).
"Promotes" certain error log messages to critical.
When alert manager's response is http, the header is removed before logging it.
Fixes minor bug (confusion calling a method that takes low/hi with params low/size ).

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 68 / 85 (80.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|2|6|33.33%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|4|8|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|51|57|89.47%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 75 / 86 (87.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 75 / 86 (87.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

File: core/src/main/java/io/questdb/ServerMain.java [hunk 1]
```diff
@@ -212,6 +212,7 @@ public ServerMain(String[] args) throws Exception {
 
             System.gc();
 
+            log.critical().$(buildInformation).$();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/ServerMain.java:215]: debris ?

File: core/src/main/java/io/questdb/log/LogAlertSocket.java [hunk 2]
```diff
@@ -230,9 +234,9 @@ public boolean send(int len, Consumer<String> ackReceiver) {
                     .$("None of the configured alert managers are accepting alerts.\n")
                     .$("Giving up sending after ")
                     .$(maxSendAttempts)
-                    .$(" attempts: ")
-                    .$(Chars.stringFromUtf8Bytes(outBufferPtr, len))
-                    .$();
+                    .$(" attempts: [")
+                    .$(Chars.stringFromUtf8Bytes(outBufferPtr, outBufferPtr + len))
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogAlertSocket.java:238]: could we add a function to the Logger that is able to log utf8 bytes without creating a string?

File: core/src/main/java/io/questdb/log/LogAlertSocket.java [hunk 3]
```diff
@@ -304,6 +308,69 @@ int getDefaultAlertPort() {
         return defaultPort;
     }
 
+
+    private static boolean isContentLength(CharSequence message, int lo, int hi) {
+        int len = hi - lo;
+        if (len < CONTENT_LENGTH.length) {
+            return false;
+        }
+        for (int i = 0; i < CONTENT_LENGTH.length; i++) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogAlertSocket.java:317]: There is faster and better way to check for constants in CharSequence. Please have a look at `SqlKeywords`:

```java
    public static boolean isAlterKeyword(CharSequence tok) {
        if (tok.length() != 5) {
            return false;
        }

        int i = 0;
        return (tok.charAt(i++) | 32) == 'a'
                && (tok.charAt(i++) | 32) == 'l'
                && (tok.charAt(i++) | 32) == 't'
                && (tok.charAt(i++) | 32) == 'e'
                && (tok.charAt(i) | 32) == 'r';
    }

```

File: core/src/main/java/io/questdb/log/LogAlertSocket.java [hunk 4]
```diff
@@ -304,6 +308,69 @@ int getDefaultAlertPort() {
...
+            if ((message.charAt(lo + i) | 32) != CONTENT_LENGTH[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @VisibleForTesting
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/log/LogAlertSocket.java:325]: there is already available IntelliJ annotation `@TestOnly`, we could perhaps standardise on that? We already use their `@Nullable` etc, may as well use the one above

File: core/src/main/java/io/questdb/std/BytecodeAssembler.java [hunk 5]
```diff
@@ -189,7 +189,7 @@ public void endMethod() {
     public void endMethodCode() {
         int len = position() - codeStart;
         if (len > 64 * 1024) {
-            LOG.error().$("Too much input to generate ").$(host.getName()).$(". Bytecode is too long").$();
+            LOG.critical().$("Too much input to generate ").$(host.getName()).$(". Bytecode is too long").$();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/BytecodeAssembler.java:192]: not critical

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by bluestreak01** [comment]
which bug is fixed?

File: core/src/main/java/io/questdb/ServerMain.java (hunk 1 above)
Comment (marregui) [reply core/src/main/java/io/questdb/ServerMain.java:215]: Now that it is pointed out, yes. I thought I'd send a ping to the alert manager at this point in the start up process, as confirmation that the server is running. But looks hacky, will remove.

File: core/src/main/java/io/questdb/log/LogAlertSocket.java (hunk 3 above)
Comment (marregui) [reply core/src/main/java/io/questdb/log/LogAlertSocket.java:317]: yeah, that is where I took inspiration from, my solution does not unroll, I considered it but it was far less readable. I will change it.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 93 (86.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|52|55|94.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 90 / 102 (88.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|8|10|80.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|59|60|98.33%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 102 (85.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|10|40.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|59|60|98.33%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 104 / 117 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|10|40.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|62|63|98.41%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 115 / 132 (87.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|10|40.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|30|34|88.24%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|62|63|98.41%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 118 / 132 (89.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|8|10|80.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|30|34|88.24%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|62|63|98.41%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 115 / 132 (87.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|10|40.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|30|34|88.24%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|62|63|98.41%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|10|10|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by bluestreak01** [comment]
thank you Miguel 👍 having `critical` level will help a lot of people

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 118 / 132 (89.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/log/Logger.java|8|10|80.00%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|30|34|88.24%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|62|63|98.41%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 115 / 132 (87.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ServerMain.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/std/BytecodeAssembler.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/log/Logger.java|4|10|40.00%|
|:large_blue_circle:|io/questdb/log/NullLogRecord.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/BuildInformationHolder.java|30|34|88.24%|
|:large_blue_circle:|io/questdb/log/LogAlertSocket.java|62|63|98.41%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogAlertSocketWriter.java|10|10|100.00%|

**Pull request merged**